File: android/src/main/kotlin/com/sparkleo/flutter_blue_background/flutter_blue_background/CharacteristicValueDispatcher.kt
package com.sparkleo.flutter_blue_background.flutter_blue_background

import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.EventChannel

/**
 * Streams GATT characteristic value events (read, write, notification) to Dart.
 */
object CharacteristicValueDispatcher {

    private val mainHandler = Handler(Looper.getMainLooper())
    private val lock = Any()

    @Volatile
    private var eventSink: EventChannel.EventSink? = null

    @Volatile
    private var sinkGeneration = 0

    private val cache = ArrayList<Map<String, Any?>>()
    private const val MAX_CACHE = 64

    fun setSink(sink: EventChannel.EventSink?) {
        if (sink == null) {
            eventSink = null
            sinkGeneration++
            return
        }
        eventSink = sink
        val generation = sinkGeneration
        val snapshot = synchronized(lock) { cache.toList() }
        mainHandler.post {
            if (eventSink === sink && generation == sinkGeneration) {
                snapshot.forEach { sink.success(it) }
            }
        }
    }

    fun emit(event: Map<String, Any?>) {
        synchronized(lock) {
            cache.add(event)
            while (cache.size > MAX_CACHE) {
                cache.removeAt(0)
            }
        }
        val sink = eventSink ?: return
        val generation = sinkGeneration
        mainHandler.post {
            if (eventSink === sink && generation == sinkGeneration) {
                sink.success(event)
            }
        }
    }

    fun reset() = synchronized(lock) { cache.clear() }
}

File: ios/flutter_blue_background/Sources/flutter_blue_background/CharacteristicValueDispatcher.swift
import Foundation
import Flutter

enum CharacteristicValueDispatcher {
    private static let maxCache = 64
    private static var cache: [[String: Any]] = []
    private static var eventSink: FlutterEventSink?
    private static var sinkGeneration = 0

    static func setSink(_ sink: FlutterEventSink?) {
        if sink == nil {
            eventSink = nil
            sinkGeneration += 1
            return
        }
        eventSink = sink
        let generation = sinkGeneration
        let snapshot = cache
        DispatchQueue.main.async {
            guard eventSink != nil, generation == sinkGeneration else { return }
            snapshot.forEach { sink?($0) }
        }
    }

    static func emit(_ event: [String: Any]) {
        cache.append(event)
        while cache.count > maxCache {
            cache.removeFirst()
        }
        guard let sink = eventSink else { return }
        let generation = sinkGeneration
        DispatchQueue.main.async {
            guard eventSink != nil, generation == sinkGeneration else { return }
            sink(event)
        }
    }

    static func reset() {
        cache.removeAll()
    }
}
